Implement GID security for the GlusterFS dynamic provisioner.

obnoxxx opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:

This PR implements GID security for the glusterfs dynamic provisioner.
It is a reworked version of PR #37549 .

<!--
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
-->

**Release note**:

<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
The glusterfs dynamic volume provisioner will now choose a unique GID for new persistent volumes from a range that can be configured in the storage class with the "gidMin" and "gidMax" parameters. The default range is 2000 - 2147483647 (max int32).
```

humblec committed:
glusterfs: teach provisioner to extract gid-range from storage class

obnoxxx committed:
glusterfs: implement GID security in the dynamic provisioner

Signed-off-by: Michael Adam <obnox@redhat.com>

obnoxxx commented:
This is an RFC on this reworked version.
It will need testing and probably a few fixes.

obnoxxx commented:
@liggitt @humblec @jsafrane - FYI

liggitt commented on minimax.go:
```diff
@@ -0,0 +1,150 @@
+// for high absolute allocation number with a bitmap.
+//
+
+package allocator
```
I'd put this under the gluster package for now, change the package name, fix up the boilerplate, and package-qualify the `allocator.Interface` reference

humblec replied:
Done

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -711,5 +739,18 @@ func parseClassParameters(params map[string]string, kubeClient clientset.Interfa
+		cfg.gidMax = defaultGidMax
+	}
+
+	if cfg.gidMin >= cfg.gidMax {
```
`if cfg.gidMin > cfg.gidMax`

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -454,6 +540,23 @@ func (d *glusterfsVolumeDeleter) Delete() error {
+		return fmt.Errorf("glusterfs: failed to parse the gid")
+	}
+
+	gidTable, err := d.plugin.getGidTable(class)
```
we've already parsed the cfg, pass in the class name, min, and max

humblec replied:
Done.

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -454,6 +540,23 @@ func (d *glusterfsVolumeDeleter) Delete() error {
 
 	glog.V(4).Infof("glusterfs: deleting volume %q with configuration %+v", volumeId, d.provisioningConfig)
 
+	gid, err := d.getGid()
```
if deleting an old one, it won't have a gid, right? we should tolerate missing ones

obnoxxx replied:
ok. right.
do we ignore non-existence of the gid or any failure?

obnoxxx replied:
done. ignoring only if not present. on real error aborting

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -499,15 +602,39 @@ func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
 	glog.V(4).Infof("glusterfs: creating volume with configuration %+v", r.provisioningConfig)
-	glusterfs, sizeGB, err := r.CreateVolume()
+
+	gidTable, err := r.plugin.getGidTable(class)
```
same here, getGidTable only needs the name, min, and max

humblec replied:
Done.

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -499,15 +602,39 @@ func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
+
+	glusterfs, sizeGB, err := r.CreateVolume(gid)
 	if err != nil {
+		err = gidTable.Release(gid)
```
don't overwrite `err`, do
```
if releaseErr := gidTable.Release(gid); releaseErr != nil {
  ... log
}
```

humblec replied:
@obnoxxx this need to be corrected. we are overwriting `err` and its not good.

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -436,6 +439,89 @@ func (d *glusterfsVolumeDeleter) GetPath() string {
+	}
+
+	if p.gidTable[class.name] == nil {
+		p.gidTable[class.name], err = NewMinMaxAllocator(defaultGidMin, defaultGidMax)
```
needs to be the absolute min/max (0-uint32max), not just the defaults

obnoxxx replied:
done

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -436,6 +439,89 @@ func (d *glusterfsVolumeDeleter) GetPath() string {
+		return err
+	}
+
+	if p.gidTable[class.name] == nil {
```
getGIDTable needs to be threadsafe, so guard gidTable with a lock, do a fast check under lock, build if needed, then do a follow up check under lock before returning

```
p.gidTableLock.Lock()
allocator, ok := p.gidTable[className]
p.gidTableLock.Unlock()
if ok {
  allocator.SetMinMax(min, max)
  return allocator, nil
}

// build, collect, set min/max ...
newAllocator := ... 

p.gidTableLock.Lock()
defer p.gidTableLock.Unlock()
// if one showed up since we last checked, use it
if allocator, ok := p.gidTable[className]; ok {
  allocator.SetMinMax(min, max)
  return allocator, nil
}
// otherwise, use the one we built
p.gidTable[className] = newAllocator
return newAllocator, nil
```

obnoxxx replied:
done thanks

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -436,6 +439,89 @@ func (d *glusterfsVolumeDeleter) GetPath() string {
+// Traverse the PVs, fetching all the GIDs from those
+// in a given storage class, and mark them in the table.
+//
+func (p *glusterfsPlugin) collectGids(class *storage.StorageClass) error {
```
pass in the storage class name and the allocator, no access to gidTable should be needed

obnoxxx replied:
no need to pass in the min/max:
when this is called, we have the full range:
it is only called when the gidTable is newly instantiated, i.e. before resetting the range to the SC values

obnoxxx replied:
className change --> done

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -436,6 +439,89 @@ func (d *glusterfsVolumeDeleter) GetPath() string {
+	gidStrings = []
+
+	for _, pv := range pvList.Items {
+		if pv.Annotations[annDynamicallyProvisioned] != glusterfsPluginName {
```
I don't think we care about this... if a PV is tied to the storage class and has a gid, we want to add it

obnoxxx replied:
agreed. done

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -436,6 +439,89 @@ func (d *glusterfsVolumeDeleter) GetPath() string {
+			continue
+		}
+
+		gidStr := pv.Annotations[volumehelper.VolumeGidAnnotationKey]
```
old ones will be missing gid strings, skip them

obnoxxx replied:
done

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -436,6 +439,89 @@ func (d *glusterfsVolumeDeleter) GetPath() string {
+			continue
+		}
+
+		if pv.Annotations[annStorageClass] != class.name {
```
use `util.GetVolumeStorageClass(pv)`

obnoxxx replied:
done

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -67,6 +68,8 @@ const (
 	defaultGidMin               = 2000
 	defaultGidMax               = math.MaxUint32
+	annDynamicallyProvisioned   = "pv.kubernetes.io/provisioned-by"
+	annStorageClass             = "volume.beta.kubernetes.io/storage-class"
```
don't need these, use use `util.GetVolumeStorageClass(pv)` to extract the class name, and I don't think we care who provisioned the pv, only whether it has an associated gid

obnoxxx replied:
done

obnoxxx replied:
also used util.GetClaimStorageClass() in another place

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -436,6 +439,89 @@ func (d *glusterfsVolumeDeleter) GetPath() string {
+	}
+
+	var gidStrings []string
+	gidStrings = []
```
unused

humblec replied:
Done.

obnoxxx commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -436,6 +439,89 @@ func (d *glusterfsVolumeDeleter) GetPath() string {
+			return err
+		}
+
+		err = p.gidTable[class.name].Allocate(gid)
```
ok. just warning in case of already allocated.

obnoxxx commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -436,6 +439,89 @@ func (d *glusterfsVolumeDeleter) GetPath() string {
+		gid, err := convertGid(gidStr)
+		if err != nil {
+			glog.Errorf("glusterfs: failed to convert gid %v", gidStr)
+			return err
```
hmm. yeah, maybe just logging would be good. done

liggitt replied:
repush?

liggitt replied:
should be this instead?
```
glog.Error(err)
continue
```

obnoxxx commented:
Update amended patchset addressing the above requests.
@liggitt Note that I have done quite changes to the minmax.go and hence put it under my authorship for now. If you would like to author it, i'll happily revert it back!

obnoxxx commented:
updated with build fix

obnoxxx commented:
still fixing a few compile errors...

obnoxxx commented:
amended with more build fixes. almost there.

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -420,6 +443,122 @@ func (d *glusterfsVolumeDeleter) GetPath() string {
+	gidStr, ok := d.spec.Annotations[volumehelper.VolumeGidAnnotationKey]
+
+	if !ok {
+		return 0, ErrNotFound
```
I'm not actually sure a missing one is an error... maybe return `gid uint32, exists bool, err error` and return `0,false,nil` in the missing case... that simplifies the calling code as well

obnoxxx replied:
for our purposes it is a valid return message (Not Found). I think the exists thing is overkill here and makes the caller even more complicated (i though). But if you insist, I'll do it ... :-)

obnoxxx replied:
got it. done

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -438,6 +577,24 @@ func (d *glusterfsVolumeDeleter) Delete() error {
+	gid, err := d.getGid()
+	if err != nil && err != ErrNotFound {
+		glog.Errorf("glusterfs: failed to parse the gid")
+		return fmt.Errorf("glusterfs: failed to parse the gid")
```
what does recovery look like in this situation? they manually fix up the annotation on the PV and we try to delete it later? are we ok blocking deletion until that happens? wondering if we should just log or call util.HandleError() and continue deleting

humblec replied:
IMO,  we need to continue and delete the volume and log the error.

obnoxxx replied:
ok. probably true. delete nonetheless and possibly sacrifice a gid? next restart it'll be fixed

obnoxxx replied:
done

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -438,6 +577,24 @@ func (d *glusterfsVolumeDeleter) Delete() error {
+
+		err = gidTable.Release(int(gid))
+		if err != nil {
+			return fmt.Errorf("glusterfs: failed to release gid: %s", err)
```
include the gid that failed

obnoxxx commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -420,6 +443,122 @@ func (d *glusterfsVolumeDeleter) GetPath() string {
+		return 0, ErrNotFound
+	}
+
+	gid, err := convertGid(gidStr)
```
sorry, not entirely sure what it means "wrapping the error".

liggitt replied:
if the annotation is malformed, the code currently logs about three times

convertGid(), line 408:
glog.Errorf("glusterfs: failed to parse gid %v ", inputGid)

getGid(), line 555:
glog.Errorf("glusterfs: failed to parse gid %v", gid)

Delete(), line 582:
glog.Errorf("glusterfs: failed to parse the gid")

instead, convertGid() should return an error like this, without logging:
```
return 0, fmt.Errorf("glusterfs: failed to parse gid %v: %v", inputGid, err)
```

getGid() can then just do this, also without logging:
```
gid, err := convertGid(gidStr)
return gid, true, err
```

and Delete() can log the result without wrapping the error:
```
if gid, exists, err := d.getGid(); err != nil {
  // being unable to get the gid from the PV should not stop deletion of it
  glog.Error(err)
} else if exists {
  gidTable, err := d.plugin.getGidTable(class.Name, cfg.gidMin, cfg.gidMax)
  if err != nil {
    return err
  }
  if err := gidTable.Release(int(gid)); if err != nil {
    return err
  }
}
```

liggitt replied:
still outstanding

obnoxxx commented:
updated addressing the comments where i commented done.

obnoxxx commented:
@liggitt i re-authered the minmax to you. jfyi. I thought it would be better for you to have to opt out than in :-)

liggitt commented:
@k8s-bot ok to test

humblec commented:
@obnoxxx can you include this https://github.com/humblec/kubernetes/commit/20beadd947eaf3180b9f27526bffbe00724894f7 in this patch set ?

obnoxxx commented:
@humblec thanks, will incorporate and re-push

obnoxxx commented:
fixed the jenkins verification error (by incorporating @humblec 's gofmt fixes).

@liggitt  can you retrigger the test?

obnoxxx commented:
fixed one more jenkins failure... (%s was used to print an int in debug ...)

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -420,6 +443,122 @@ func (d *glusterfsVolumeDeleter) GetPath() string {
+		gidStr, ok := pv.Annotations[volumehelper.VolumeGidAnnotationKey]
+
+		if !ok {
+			glog.Errorf("glusterfs: no gid found in pv")
```
lower log level to info, probably, and include PV name if we want this to be actionable

obnoxxx replied:
done

obnoxxx commented:
accidentially pushed a fix i needed to run local-up-cluster. overpushed without it...

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -420,6 +443,122 @@ func (d *glusterfsVolumeDeleter) GetPath() string {
+
+		_, err = gidTable.Allocate(int(gid))
+		if err == ErrConflict {
+			glog.Errorf("glusterfs: warnng: gid %v already allocated", gid)
```
use `glog.Warningf`, drop the "warning" text from the message, and include the pv name

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -482,16 +638,35 @@ func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
+	glusterfs, sizeGB, err := r.CreateVolume(gid)
 	if err != nil {
+		_err := gidTable.Release(gid)
+		if _err != nil {
```
`if releaseErr := gidTable.Release(gid); releaseErr != nil {`

obnoxxx replied:
I personally don't like putting the assignment and the check into one line (coming from C programming and debugging....). but I can rename the variable.

obnoxxx replied:
done (renamed variable)

obnoxxx replied:
Ok... giving in... ;-) This seems to be so common here... moved it into one line.

obnoxxx replied:
done

liggitt commented on pkg/volume/glusterfs/glusterfs_minmax.go:
```diff
@@ -0,0 +1,174 @@
+	}
+
+	// inconsistency: no free items, but free > 0. fix that
+	a.free = 0
```
if we ever hit this, this type has a bug, right? I'd rather not do this, and make sure our test coverage is good

obnoxxx replied:
done

liggitt commented on pkg/volume/glusterfs/glusterfs_minmax.go:
```diff
@@ -0,0 +1,174 @@
+var ErrOutOfRange = errors.New("out of range")
+var ErrRangeFull = errors.New("range full")
+
+type MinMaxAllocator struct {
```
definitely need tests to exercise this

obnoxxx replied:
wrote testsuite. pushing shortly...

obnoxxx replied:
done

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37886)
<!-- Reviewable:end -->

obnoxxx commented:
Wrote an extensive testsuite for the minmax allocator.
Fixed the requested bits.
Also fixed an error in the minmax allocatore found by the test. :-)

obnoxxx commented:
updated with changes for last requests by @liggitt

obnoxxx commented:
I had missed to address the request for the return type and logging in getGid, convertGid and friends. done now.

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -420,6 +442,120 @@ func (d *glusterfsVolumeDeleter) GetPath() string {
+		gid, err := convertGid(gidStr)
+		if err != nil {
+			glog.Error(err)
+			return err
```
just log and continue if we can't get a valid gid from an existing PV

obnoxxx committed:
glusterfs: add MinMaxAllocator

An allocator of integers that allows for changing the range.
Previously allocated numbers are not lost, and can  be
released later even if they have fallen outside of the range.

Signed-off-by: Michael Adam <obnox@redhat.com>

liggitt commented:
/lgtm

liggitt commented:
@humblec @jsafrane - any last comments?

humblec commented:
@liggitt @jsafrane just to mention, if there are statically provisioned ( with `gid` annot. ) PVs, the `same GIDs may` get dispatched from our allocator, depends on the range.

liggitt commented:
I don't think our provisioner should be concerned with PVs outside its storage class when determining available GIDs

childsb commented:
@k8s-bot kops aws e2e test this

pwittrock commented:
I am seeing this failure when I try to do `make release`.  Does that look related to this PR?

```sh
# k8s.io/kubernetes/pkg/volume/glusterfs
pkg/volume/glusterfs/glusterfs.go:510: constant 4294967295 overflows int
+++ [1205 22:54:42] linux/arm64: go build started
+++ [1205 22:58:45] linux/arm64: go build finished
Makefile:79: recipe for target 'all' failed
make[1]: *** [all] Error 1
Makefile:264: recipe for target 'cross' failed
make: *** [cross] Error 1
Makefile:248: recipe for target 'release' failed
```

liggitt commented:
Yes, see #37886 to fix the cross compile for 32 bit

liggitt commented:
Oops, #38137
